src/controller/domain_handle.rs

```diff
@@ -34,6 +34,7 @@ pub struct DomainHandle {
 
     cr_poll: PollingLoop<ControlReplyPacket>,
     shards: Vec<DomainShardHandle>,
+    nodes: Vec<NodeIndex>,
```

Comment:
It isn't clear whether we really want to track nodes here. I'm afraid it may be too easy for it to go out of sync with the controller's state.

Reply:
I agree, but it's tricky to get access to the controller state from within a `DomainHandle`.

Maybe we could make `assignments` available to the controller via an API, but that would still require the controller to be able to map the domain to its nodes. Is such a mapping available? (@jonhoo)

Reply:
I implemented a possible fix in 926f0f28bb6d77d4c828966257aa3c1163350343, keeping state only on the controller side. This has the effect of making recovery O(|D| * |V|), where D is the set of domains, and V is the set of data-flow vertices, however. This could be solved by keeping separate vertex->domain assignment state in `ControllerInner`.